### DISCUSSION TITLE
List of encoder supported

### DISCUSSION THREAD

**Discussion by dayuyang1999** [castorini/pyserini#1950] [Q&A] [answered]
Hi Team,

Can I use any embedding model for creating dense index?

If not, is there a list of supported models?

Best,
Dayu

**Comment by lintool** [comment]
Yes, start here: https://github.com/castorini/pyserini/blob/master/docs/usage-index.md#building-a-dense-vector-index

I would also suggest that you go through our onboarding guide: https://github.com/castorini/onboarding/blob/master/README.md#-foundations-of-retrieval

**Reply by dayuyang1999** [reply]
Hi,

Thanks for the response. 

I ready read those contents, but still they didn't provide a clear list of supported embedding model.

Specifically, I was trying to use e5-mistral-7b-instruct for embedding. Is that supported (since it is a instruct model so there suppose to have a natural language instruct input but I didn't see any place I can input that when creating FaissSearcher)

**Reply by lintool** [reply]
Pyserini should be able to work with any embedding model on HuggingFace, by just pointing at it. For instructions, look at the BGE instructions [1].

[1] https://castorini.github.io/pyserini/2cr/msmarco-v1-passage.html

**Reply by dayuyang1999** [reply]
Hi,

Thanks for the reply.

The link you sent to me seems to be a leaderboard rather than an instruction?

Would you mind double check the link?

**Reply by lintool** [reply]
Scroll down until you get to "BGE-base-en-v1.5: PyTorch", expand the tab, you should get instructions on how to run. You should be able to plug in your own models based on those instructions. Look at the code as appropriate.

**Reply by dayuyang1999** [reply]
Hi,

Thanks for prompt reply.

The link you provided gave the command to generate run on TREC 2019 queries:

```bash
python -m pyserini.search.faiss \
  --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder BAAI/bge-base-en-v1.5 --l2-norm --query-prefix "Represent this sentence for searching relevant passages:" \
  --index msmarco-v1-passage.bge-base-en-v1.5 \
  --topics dl19-passage \
  --output run.msmarco-v1-passage.bge-base-en-v1.5-pytorch.dl19.txt \
  --hits 1000


```


However, I found that, for creating dense index, there is no `--query-prefix` argument



https://github.com/castorini/pyserini/blob/master/pyserini/encode/__main__.py

**Reply by lintool** [reply]
I believe @MXueguang did the encoding using Tevatron [1] - but if you want to add the feature to `pyserini.encode`, pull requests welcome!

[1] http://tevatron.ai/

**Reply by dayuyang1999** [reply]
Thank you so much! One more question:

According to the training script in tevatron, does that mean the model trained using default similarity measure used in tevatron is optimized on `dot product` instead of `cosine similarity`?

So using a model trained using tevatron to compute cosine similarity is wrong?

`https://github.com/texttron/tevatron/blob/main/src/tevatron/retriever/modeling/encoder.py`

**Reply by dayuyang1999** [reply, marked as answer]
Oh seems not.

if I need a model to produce cosine similarity, I can do that by passing --normalize = True
